File: 0935-knight-dialer/0935-knight-dialer.cpp
const int MOD = 1e9 + 7;

class Solution {
public:
    unordered_map<int, vector<int>> moves = {
        { 0, { 4, 6 } },
        { 1, { 6, 8 } },
        { 2, { 7, 9 } },
        { 3, { 4, 8 } },
        { 4, { 0, 3, 9 } },
        { 5, {} },
        { 6, { 0, 1, 7 } },
        { 7, { 2, 6 } },
        { 8, { 1, 3 } },
        { 9, { 2, 4 } }
    };
    
    int knightDialer(int n) {
        vector<int> dp(10, 1);
        for (int len = 2; len <= n; len++) {
            vector<int> dpCopy = dp;
            for (int button = 0; button < 10; button++) {
                dp[button] = 0;
                for (auto it: moves[button]) {
                    dp[button] = (dp[button] + dpCopy[it]) % MOD;
                }
            }
        }
        int res = 0;
        for (int button = 0; button < 10; button++) {
            res = (res + dp[button]) % MOD;
        }
        return res;
    }
};

File: 0935-knight-dialer/0935-knight-dialer.java
class Solution {
    public static final int MOD = 1000000007;
    
    public int knightDialer(int n) {
        int[] dp = new int[10];
        for (int button = 0; button < 10; button++) {
            dp[button] = 1;
        }
        for (int len = 2; len <= n; len++) {
            int[] dpCopy = new int[10];
            dpCopy[0] = (dp[4] + dp[6]) % MOD;
            dpCopy[1] = (dp[6] + dp[8]) % MOD;
            dpCopy[2] = (dp[7] + dp[9]) % MOD;
            dpCopy[3] = (dp[4] + dp[8]) % MOD;
            dpCopy[4] = ((dp[0] + dp[3]) % MOD + dp[9]) % MOD;
            dpCopy[5] = 0;
            dpCopy[6] = ((dp[0] + dp[1]) % MOD + dp[7]) % MOD;
            dpCopy[7] = (dp[2] + dp[6]) % MOD;
            dpCopy[8] = (dp[1] + dp[3]) % MOD;
            dpCopy[9] = (dp[2] + dp[4]) % MOD;
            dp = dpCopy;
        }
        int res = 0;
        for (int button = 0; button < 10; button++) {
            res = (res + dp[button]) % MOD;
        }
        return res;
    }
}